### PR TITLE
HTCONDOR-2114 periodic-vacate

### PR DESCRIPTION
**Pull request by GregThain**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [ ] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [ ] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed  (documentation [build logs](https://readthedocs.org/projects/htcondor/builds/))
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

**Comment by GregThain** [comment]
```mermaid
.. mermaid::
   :caption: Machine states and the possible transitions between the states
   :align: center

   stateDiagram-v2
     direction LR
     [*]--> Owner
     Owner --> Unclaimed: A
     Unclaimed --> Matched: C
     Unclaimed --> Owner: B
     Unclaimed --> Drained: P
     Unclaimed --> Backfill: E
     Unclaimed --> Claimed: D
     Backfill  --> Owner: K
     Backfill  --> Matched: L
     Backfill  --> Claimed: M
     Matched --> Claimed: G
     Matched --> Owner: F
     Claimed --> Preempting: H
     Preempting --> Owner: J
     Preempting --> Claimed: I
     Owner --> Drained: N
     Drained --> Owner: O

```

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
The periodic vacate is evaluated at on-exit time (like the other periodic exprs are). Do we want that and if so, what should it mean?
The schedd, job router, and gridmanager evaluate job policy. They need to updated to handle VACATE_FROM_RUNNING. The job router and gridmanager can't vacate a running job, so we need to decide what they will do.
For Condor-C jobs, the gridmanager removes all of the policy attributes for the downstream job it submits. Should it do the same for vacate?

File: src/condor_utils/user_job_policy.cpp [hunk 1]
```diff
@@ -310,6 +310,14 @@ UserPolicy::AnalyzePolicy(ClassAd & ad, int mode, int state)
 				return HOLD_IN_QUEUE;
 			}
 		}
+
+		/* Should I perform a periodic vacate? */
+		if ((mode == PERIODIC_ONLY) && ((state == RUNNING) || (state == SUSPENDED))) {
```
Comment (JaimeFrey) [inline src/condor_utils/user_job_policy.cpp:315]: The check of state is redundant here with the enclosing if().

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
The job router's policy evaluation still needs updating.

File: src/condor_job_router/JobRouter.cpp [hunk 2]
```diff
@@ -1037,6 +1037,9 @@ JobRouter::EvalSrcJobPeriodicExpr(RoutedJob* job)
 			// the job as if it is a new job from the schedd.
 			ret_val = true;
 			break;
+		case VACATE_FROM_RUNNING:
+			SetJobIdle(job);
```
Comment (JaimeFrey) [inline src/condor_job_router/JobRouter.cpp:1041]: SetJobIdle() just sets the JobStatus of the source job to Idle. I think the Job Router should treat VACATE_FROM_RUNNING the same as STAYS_IN_QUEUE (i.e. do nothing). The job policy expressions are forwarded to the routed job, where PeriodicVacate can have its intended effect. Evaluating the policy expressions in both the unrouted and routed jobs is highly suspect to me, but that's an issue for another ticket.

**Review by JaimeFrey — CHANGES_REQUESTED** [review]

**Review by JaimeFrey — APPROVED** [review]
LGTM

**Pull request merged**